### PR TITLE
[Feature] option to delete a data source 🚀 

### LINKED ISSUE #1554: Option to remove added datasource
### Is your proposal related to a problem?

I added the same datasource (Google Sheets) twice and I wanted to remove one but there is no option.

Screenshot 2021-12-09 at 7 19 45 PM

### Describe the solution you'd like

A delete/remove button next to the list of added datasources.

### PR DESCRIPTION
**Pull request by shah21**
Resolves #1554

### CODE REVIEW

**Comment by akshaysasidrn** [comment]
@shah21 Please add a spec with app having multiple versions with data sources and data queries such that deleting data source in a particular version only deletes the corresponding data queries.

**Comment by shah21** [comment]
> @shah21 Please add a spec with app having multiple versions with data sources and data queries such that deleting data source in a particular version only deletes the corresponding data queries.

will do that

File: server/src/controllers/data_sources.controller.ts [hunk 1]
```diff
@@ -74,6 +85,26 @@ export class DataSourcesController {
...
+    const ability = await this.appsAbilityFactory.appsActions(req.user, {
+      id: app.id,
+    });
+
+    if (!ability.can('deleteQuery', dataSource.app)) {
+      throw new ForbiddenException('you do not have permissions to perform this action');
+    }
+
+    const result = await this.dataSourcesService.delete(params.id);
+    return decamelizeKeys(result);
```
Comment (gsmithun4) [inline server/src/controllers/data_sources.controller.ts:105]: Do we need to send back postgres response back? shall we check if the operation is success from response and send back empty response with status `200`?

**Review by sherfin94 — CHANGES_REQUESTED** [review]
image

The new delete button seem to be breaking down the label of certain data sources. Maybe increase the width of the popover a bit?

Other than this everything looks fine editor-wise.

File: server/src/controllers/data_sources.controller.ts [hunk 2]
```diff
@@ -81,6 +94,28 @@ export class DataSourcesController {
...
+    const dataSourceId = params.id;
+
+    const dataSource = await this.dataSourcesService.findOne(dataSourceId);
+
+    const app = await this.appsService.find(dataSource.appId);
+    const ability = await this.appsAbilityFactory.appsActions(req.user, {
+      id: app.id,
+    });
+
+    if (!ability.can('deleteQuery', dataSource.app)) {
```
Comment (sherfin94) [inline server/src/controllers/data_sources.controller.ts:109]: Can we add a `deleteSource` ability type and use it?

**Review by sherfin94 — CHANGES_REQUESTED** [review]
Requested one change, functionality works perfectly!

File: frontend/src/Editor/LeftSidebar/SidebarDatasources.jsx [hunk 3]
```diff
@@ -25,14 +67,39 @@ export const LeftSidebarDataSources = ({ appId, editingVersionId, darkMode, data
...
+            {dataSource.name}
+          </span>
+        </div>
+        <div className="col-auto">
+          <button
+            className="btn btn-sm ds-delete-btn"
+            onClick={() => deleteDataSource(dataSource)}
+            style={
+              {
+                // display: this.state.showHiddenOptionsForDataQueryId === dataQuery.id ? 'block' : 'none',
```
Comment (gsmithun4) [inline frontend/src/Editor/LeftSidebar/SidebarDatasources.jsx:80]: Remove the commented code

**Review by gsmithun4 — APPROVED** [review]
LGTM

**Review by akshaysasidrn — DISMISSED** [review]

**Comment by Nandinisaha13** [comment]
<html>
<body>
<google-sheets-html-origin>Test Cases

Test Case | Steps | Result
-- | -- | --
Delete button should work for added data sources. | Create an application. | Application gets created.
  | Add data source & save it. | Data sources should be added.
  | Click on the data source. | It should have delete button on the right corner.
  | Click on the delete button. | It gets pop up
Delete should delete the queries created corresponding it. | Create an application. | Application gets created.
  | Add data source & save it. | Data sources should be added.
  | Create a query for the added data source. | The query gets created.
  | Click on the data source. | It should have delete button on the right corner.
  | Click on the delete button. | It deletes the datasource and the query
Imported application should have data source delete button as working. | Import the application. | The application gets imported.
  | Click on datasource. | Data sources is displayed.
  | Click on delete button. | It deletes the datasource & query with it.


</body>
</html>

**Review by gsmithun4 — CHANGES_REQUESTED** [review]
Found two issues
1. Delete conformation alert box style (button should be red)
2. https://www.loom.com/share/e20f911fa6844f588f34c21dd65c0532

**Comment by shah21** [comment]
> Found two issues
> 
> 1. Delete conformation alert box style (button should be red)
> 2. https://www.loom.com/share/e20f911fa6844f588f34c21dd65c0532

The second one is a general issue.its not related to my changes. It existed in develop branch for a long. something goes wrong after coming back to same data source modal

**Review by gsmithun4 — APPROVED** [review]
LGTM

**Review by sherfin94 — APPROVED** [review]

**Pull request merged**